straxen/plugins/peak_shadow.py

```diff
@@ -0,0 +1,79 @@
+import numpy as np
+import strax
+import numba
+export, __all__ = strax.exporter()
+
+
+@export
+@strax.takes_config(
+    strax.Option('pre_s2_area_threshold', default=1000,
+                 help='Only take S2s large than this into account'
+                 ' when calculating PeakShadow [PE]'),
+    strax.Option('time_window_backward', default=int(3e9),
+                 help='Search for S2s causing shadow in this time window [ns]'),
+    strax.Option('skip_drift_time', default=0,
+                 help='Deduct max drift time to avoid peak interference in a single event [ns]'))
+class PeakShadow(strax.OverlapWindowPlugin):
+    """
+    This plugin can find and calculate the previous S2 shadow at peak level,
+    with time window backward and previous S2 area as options.
+    It also gives the area and position infomation of these previous S2s.
+    """
+
+    __version__ = '0.0.2'
+    depends_on = ('peak_basics', 'peak_positions')
+    provides = 'peak_shadow'
+    save_when = strax.SaveWhen.EXPLICIT
+
+    def get_window_size(self):
+        return self.config['time_window_backward']
+
+    def infer_dtype(self):
+        dtype = [('shadow', np.float32, 'previous s2 shadow [PE/ns]'),
+                 ('pre_s2_area', np.float32, 'previous s2 area [PE]'),
+                 ('shadow_dt', np.int64, 'time diffrence to the previous s2 [ns]'),
+                 ('pre_s2_x', np.float32, 'x of previous s2 peak causing shadow [cm]'),
+                 ('pre_s2_y', np.float32, 'y of previous s2 peak causing shadow [cm]')]
+        dtype += strax.time_fields
+        return dtype
+
+    def compute(self, peaks):
+        roi_shadow = np.zeros(len(peaks), dtype=strax.time_fields)
+        n_seconds = self.config['time_window_backward']
+        n_drift_time = self.config['skip_drift_time']
+        roi_shadow['time'] = peaks['center_time'] - n_seconds
+        roi_shadow['endtime'] = peaks['center_time'] - n_drift_time
+
+        mask_pre_s2 = peaks['area'] > self.config['pre_s2_area_threshold']
+        mask_pre_s2 &= peaks['type'] == 2
+        split_peaks = strax.split_touching_windows(peaks[mask_pre_s2], roi_shadow)
+        res = np.zeros(len(peaks), self.dtype)
+        res['pre_s2_x'] = np.nan
+        res['pre_s2_y'] = np.nan
+        compute_shadow(peaks, split_peaks, res)
+
+        res['time'] = peaks['time']
+        res['endtime'] = strax.endtime(peaks)
+
+        return res
+
+
+def compute_shadow(peaks, split_peaks, res):
+    if len(res):
+        return _compute_shadow(peaks, split_peaks, res)
+
+
+@numba.njit(cache=True)
+def _compute_shadow(peaks, split_peaks, res):
+    for p_i, p_a in enumerate(peaks):
+        new_shadow = 0
+        for s2_a in split_peaks[p_i]:
+            if p_a['center_time'] - s2_a['center_time'] <= 0:
```

Comment:
== 0 I can see, but < 0 should actually never happen is not it?

Reply:
I did not check whether there are <0. If the computing consumption goes the same for ==0 and <0, I think this is fine?

Reply:
Yes, I just wanted to point it out. In case you got negative values I think something might be fishy.